Draft: content and planning of building and packaging chapter (#3)
Simulation-Software-Engineering/Lecture-Material: building-and-packaging/material/exercise_packaging_text.md

- Once the output functionality has been separated from the solver, we need to bundle the solver itself into a function called `solve`. We need to do this because later on we will call this function after importing the package:

```python
from package_name import solver

solver.solve()
```

- Once you have refactored the code, try calling the `solve()` function again through a Python script as shown above or a Python shell. Compare the plots with the plots in the figure you saved earlier. Both outputs should be identical

- The function `solve()` should take in physical parameters which the user can vary. It is always a good idea to pass important physical parameters as input arguments to a function of a simulation software because you do not want to rebuild a package every time you change a parameter
- Try to think which parameters would be worth changing? These are typically related to the initial condition of the system being simulated

## Step 3 - Creating folder structure for packaging

> - Now that you have a refactored code, replicate the folder structure we learnt in the lecture to prepare the code for packaging

Reviewer: To make it a bit more comfortable here, you could link to the slides.

- With the help of the lecture notes create additional files and arrange the files in a folder structure which is similar to what is shown in the notes
- There is ambiguity in how `setup.cfg` and `setup.py` are written and it is up to you to determine how much information is relevant for the code in this exercise. Also think if you really need both `setup.cfg` and `setup.py` for this exercise? Could you manage to publish the package with just one of them?

> - The `README.md` file consists of a longer description about the code and what it does. Please try to provide as much information as possible regarding the code

Reviewer: ```suggestion
- The `README.md` file consists of a longer description about the code and what it does. Please provide some description of the code and its usage
```

## Step 4 - Create distribution packages

- With reference to the lecture notes, create distribution packages for this project

> - After creating the distribution packages, check the `dist/` folder to ensure that the archive files have been created

Reviewer: Could you describe what should be in the archive, such that I can check I did the right thing so far.

## Step 5 - Create an account and a API Token on TestPyPI

- Create an account on [TestPyPI](https://packaging.python.org/guides/using-testpypi/)
- Create an API Token on TestPyPI
- **Note down the name and token value because this is the username and password for publishing**

## Step 6 - Uploading the package

- You should already have Twine installed on your system
- Upload the distribution archives using the commands shown in the lecture notes
- Go to TestPyPI and view the package which has been uploaded
- Take a screenshot of the TestPyPI webpage which displays your package. This screenshot will be used for submission

## Step 7 - Testing the deployed package

- Using the commands from the lecture notes try to install the package using `pip` and also run the code by importing the `solve()` functionality in a Python script or an interactive Python shell

## Step 8 - Submitting the exercise

> - Open a pull request from your branch to the `main` branch of the exercise repository

Reviewer: Give name for PR
